lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,421 @@
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(pubkey_policy_tree).
+
+-include("../include/public_key.hrl").
+
+%% API
+-export([add_leaves/2,
+         add_siblings/2,
+         any_leaf/2,
+         collect_qualifiers/2,
+         constraint_policy_node_set/1,
+         empty/0,
+         in_set/2,
+         is_empty/1,
+         map_leaves/2,
+         policy_node/3,
+         prune_leaves/2,
+         prune_tree/1,
+         prune_invalid_nodes/2,
+         root/0,
+         valid_policy_node_set/1
+        ]).
+
+-export_type([policy_tree/0]).
+
+-type policy_node() ::
+        #{valid_policy := public_key:oid(),
+          qualifier_set := [#'PolicyInformation'{}],
+          expected_policy_set := [public_key:oid()]}.
+
+-type policy_tree_node()   :: {policy_node(), [policy_tree_node()] |
+                               [policy_node()]}.
+
+-opaque policy_tree()        :: {} | policy_tree_node().
+
+%%%===================================================================
+%%% Internal API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+-spec add_leaves(policy_tree(), LeafFun::function()) -> policy_tree().
+%%
+%% Add leaves specified by calling LeafFun with the current leaves
+%% as input
+%%--------------------------------------------------------------------
+add_leaves({Parent, []}, LeafFun) ->
+    {Parent, LeafFun(Parent)};
+add_leaves(Tree, LeafFun0) ->
+    LeafFun = fun(Leaf) ->
+                      NewLeaves = LeafFun0(Leaf),
+                      {Leaf, NewLeaves}
+              end,
+    map_leaves(Tree, LeafFun).
+
+%%--------------------------------------------------------------------
+-spec add_siblings(policy_tree(), SiblingFun::function()) -> policy_tree().
+%%
+%% Add sibling leaves if SiblingFun returns a list of policy nodes
+%% for the leaf parent.
+%%--------------------------------------------------------------------
+add_siblings({Parent,[{_, _}|_] = SubTree}, SiblingFun) ->
+    {Parent, lists:map(fun(ChildNode)->
+                               add_siblings(ChildNode, SiblingFun)
+                       end, SubTree)};
+add_siblings({Parent, Leaves} = Node, SiblingFun) ->
+    case SiblingFun(Parent) of
+        no_sibling ->
+            Node;
+        Siblings ->
+            {Parent, Leaves ++ Siblings}
+    end.
+
+%%--------------------------------------------------------------------
+-spec any_leaf(policy_tree(), Depth::integer()) -> no_node | [policy_node()].
+%%
+%% Find leaf policy nodes at tree <Depth>  with valid_policy = ?anyPolicy if they exists
+%%--------------------------------------------------------------------
+any_leaf({}, _) ->
+    no_node;
+any_leaf(Tree, Depth) ->
+    case depth(Tree) of
+        Depth ->
```

Comment:
- depth being compared with width(returned by depth/1) - does not make sense
- what is intention behind this depth check? which RFC part? can this be explained in code comment?

Reply:
I do not have time to finish this properly now. The intent is to check that the depth of the tree is equal to to number of certs in the path. Because of the wording in the RFC.  Wrap up part! But I am not sure if we need to check this or not. I removed the check for now and I will work on it later. Unless you find a good solution while I am away.

Reply:
I mark a comment as outdated, I think the solution is part of resolving this. Which I will be working on!

Reply:
@u3s I think I have a better solution!

Reply:
@u3s check and resolve if OK

Reply:
yes. original comment is not relevant anymore.